lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -348,44 +349,59 @@ LogicalResult static applyNoBlackBoxStyleDataTaps(const AnnoPathValue &target,
+      Operation *mod = (moduleTarget->ref.getOp());
+      FExtModuleOp extMod = dyn_cast<FExtModuleOp>(mod);
+      if (!extMod)
+        return mod->emitOpError(
+            "can specify internal_path on external modules only");
       InstanceOp modInstance;
       if (!moduleTarget->instances.empty()) {
         modInstance = moduleTarget->instances.back();
       } else {
-        auto *node = state.instancePathCache.instanceGraph.lookup(mod);
+        auto *node = state.instancePathCache.instanceGraph.lookup(
+            cast<hw::HWModuleLike>(mod));
         if (!node->hasOneUse())
-          return mod.emitOpError(
+          return mod->emitOpError(
               "cannot be used for DataTaps, it is instantiated multiple times");
         modInstance = cast<InstanceOp>((*node->uses().begin())->getInstance());
       }
       ImplicitLocOpBuilder builder(modInstance.getLoc(), modInstance);
       builder.setInsertionPointAfter(modInstance);
-      auto wireType =
-          wireTarget->ref.getOp()->getResult(0).getType().cast<FIRRTLType>();
-      // Get the InnerRef to the instance.
-      SmallVector<Attribute> symbols = {getInnerRefTo(
-          modInstance, "extModXMR", [&](FModuleOp mod) -> ModuleNamespace & {
+      auto portRefType = RefType::get(wireTarget->ref.getOp()
+                                          ->getResult(0)
+                                          .getType()
+                                          .cast<FIRRTLBaseType>());
+      StringRef refName("ref");
+      // Add RefType ports corresponding to this "internalPath" to the external
+      // module. This also updates all the instances of the external module.
+      // This removes and replaces the instance, and returns the updated
+      // instance.
+      modInstance = addPortsToModule(
+          dyn_cast<FModuleLike>(mod), modInstance, portRefType, Direction::Out,
+          refName, state.instancePathCache,
+          [&](FModuleLike mod) -> ModuleNamespace & {
             return state.getNamespace(mod);
-          })};
-      // This represents an xmr into the `moduleTarget`.
-      auto source = builder.create<VerbatimExprOp>(
-          wireType, "{{0}}." + internalPathAttr.getValue(), ValueRange{},
-          symbols);
-      source->setAttr(
-          StringAttr::get(context, "name"),
-          StringAttr::get(
-              context,
-              state.getNamespace(modInstance->getParentOfType<FModuleOp>())
-                  .newName(tapName.getValue() + "_internalPath")));
-      // Now set the xmr verbatim as the source for the final datatap xmr.
-      srcTarget = AnnoPathValue(source);
+          },
+          &state.targetCaches);
+      // The extern module can have other internal paths attached to it, append
+      // this to them.
+      auto internalPaths = extMod.getInternalPaths();
+      SmallVector<Attribute> paths(internalPaths.begin(), internalPaths.end());
```

Comment:
Could this be done without the auto either `paths(extMod.getInternalPaths())` or `paths = extMod.getInternalPath()`?